lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,111 @@
+
+#include "circt/Scheduling/Algorithms.h"
+
+#include "mlir/IR/Operation.h"
+
+#include "ortools/sat/cp_model.h"
+
+using namespace circt;
+using namespace circt::scheduling;
+using namespace operations_research;
+using namespace operations_research::sat;
+
+LogicalResult scheduling::scheduleCPSAT(SharedOperatorsProblem &prob,
+                                        Operation *lastOp) {
+  Operation *containingOp = prob.getContainingOp();
+  if (!prob.hasOperation(lastOp))
+    return containingOp->emitError("problem does not include last operation");
+
+  CpModelBuilder model;
+  auto &tasks = prob.getOperations();
+  auto &resources = prob.getOperatorTypes();
+
+  DenseMap<Operation *, IntVar> taskStarts;
+  DenseMap<Operation *, IntVar> taskEnds;
+  DenseMap<Operation *, IntervalVar> taskIntervals;
+
+  // first get horizon
+  unsigned horizon = 0;
+  for (auto *task : tasks) {
+    unsigned duration = *prob.getLatency(*prob.getLinkedOperatorType(task));
+    horizon += duration;
+  }
+
+  // build vars
+  unsigned i = 0;
+  for (auto *task : tasks) {
+    auto resource = prob.getLinkedOperatorType(task);
+    IntVar startVar = model.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = model.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    unsigned duration = *prob.getLatency(*resource);
+
+    IntervalVar taskInterval =
+        model.NewIntervalVar(startVar, duration, endVar)
+            .WithName((Twine("task_interval_") + Twine(i)).str());
+
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    taskIntervals[task] = taskInterval;
+
+    i++;
+  }
+
+  IntVar makeSpan = model.NewIntVar(Domain(0, horizon)).WithName("makespan");
+  for (const auto &task : tasks) {
+    if (task == lastOp)
+      model.AddLessOrEqual(taskEnds[task], makeSpan);
+    else {
+      for (auto *successor : task->getUsers()) {
+        model.AddLessOrEqual(taskEnds[task], taskStarts[successor]);
+      }
+    }
+  }
+
+  unsigned infinity = std::numeric_limits<unsigned>::max();
+  for (const Problem::OperatorType &resource : resources) {
+    // https://or.stackexchange.com/a/3363
+    unsigned capacity = prob.getLimit(resource).getValueOr(infinity);
+    if (capacity < infinity) {
+      CumulativeConstraint cumu = model.AddCumulative(capacity);
+      for (llvm::detail::DenseMapPair<Operation *, IntervalVar> &item :
+           taskIntervals) {
+        auto taskResource = *prob.getLinkedOperatorType(item.getFirst());
+        if (taskResource.str() == resource.str()) {
+          IntVar demandVar = model.NewIntVar(Domain(1)).WithName(
+              (Twine("demand_") + Twine(i) + Twine("_") + Twine(resource.str()))
+                  .str());
+          IntervalVar start =
+              model.NewFixedSizeIntervalVar(item.getSecond().StartExpr(), 1);
```

Comment:
conventional formulation for `SharedOperatorsProblem` (interval during which the resource is occupied has length `1`)